File: sol/solution/0001-0100/0017/Solution.rb
# @param {String} digits
# @return {String[]}
def letter_combinations(digits)
	return [] if digits.empty?
	phone = {
		'2' => %w[a b c],
		'3' => %w[d e f],
		'4' => %w[g h i],
		'5' => %w[j k l],
		'6' => %w[m n o],
		'7' => %w[p q r s],
		'8' => %w[t u v],
		'9' => %w[w x y z]
	}
	result = ['']
	digits.each_char do |digit|
		result = result.flat_map { |prefix| phone[digit].map { |char| prefix + char } }
	end
	result
end

File: sol/solution/0001-0100/0017/Solution.rs
impl Solution {
    pub fn letter_combinations(digits: String) -> Vec<String> {
        if digits.is_empty() {
            return Vec::new();
        }
        let map = vec![
            "",    // 0
            "",    // 1
            "abc", // 2
            "def", // 3
            "ghi", // 4
            "jkl", // 5
            "mno", // 6
            "pqrs",// 7
            "tuv", // 8
            "wxyz" // 9
        ];
        let mut res = vec![String::new()];
        for d in digits.chars() {
            let idx = d.to_digit(10).unwrap() as usize;
            let letters = map[idx];
            let mut tmp = Vec::new();
            for prefix in &res {
                for ch in letters.chars() {
                    let mut new_str = prefix.clone();
                    new_str.push(ch);
                    tmp.push(new_str);
                }
            }
            res = tmp;
        }
        res
    }
}